### DISCUSSION TITLE
[React] Focus management and prop getters for external reference

### DISCUSSION THREAD

**Discussion by mihkeleidast** [floating-ui/floating-ui#2717] [React]
Currently, the only way to get the correct props (both aria and interaction) and focus management is to use an "internal" reference element - i.e, the reference element must be in some context through which the floating hook return is shared. This works for lots of smaller floating components, where the reference and floating element can be co-located this way. Unfortunately, it does not work for some more complex floating elements, like modals and drawers, where there can be multiple references for a single floating element.

I put together a very simple example based on the drawer test: https://github.com/floating-ui/floating-ui/pull/2127, which should showcase the issues. If a more complex example is necessary, I can put together something where in a virtualized list (we use [react-virtuoso](https://github.com/petyosi/react-virtuoso) at the moment), each list item has a trigger that opens a drawer. The list items being rendered virtually would make even the thought of co-location of the drawer/reference impossible, as we can't have the drawer self-closing when user scrolls the list.

What I'd like to achieve, is:
- have an external reference on a floating component;
- have that external reference receive all the same props and behavior that an internal reference would get.

I guess it may be possible to do this in userland already, by syncing the hook return value to some more global state management system based on an unique ID, and then retrieving the reference props based on the same ID? I haven't gotten that far yet, wanted to file an issue anyway to get an opinion on this (should/could it be simpler?) and eventually provide some documented example for others who run into similar issues/questions.

This also sort of ties into the singleton issue (https://github.com/floating-ui/floating-ui/issues/1697), but as that is mostly described as an issue of transitions, while this is about functionality, I filed it separately. But feel free to merge these if they likely share the solution.

**Comment by atomiks** [comment]
I guess as you said, you'd need to use an effect to synchronize the values with the parent context/state, like with external elements. Or lift the hook up higher in the component tree where the values are needed, then pass its values down to the child component.

**Comment by mihkeleidast** [comment]
> Or lift the hook up higher in the component tree where the values are needed, then pass its values down to the child component.

That can be problematic if the values are passed through a simple context like is used in all of the examples, as it's rather easy to run into a conflicting scenario where you want multiple hooks/context values at a similar level, and with context, the innermost provider always wins.

But I'll see what I can do about implementing some sort of global store, and will report back in case of any issues. Or, if everything works wonderfully, we can document this.

**Comment by mihkeleidast** [comment]
@atomiks I started to investigate this topic a bit, started with looking into "lift the hook up higher in the component tree" approach. It seems like non-modal focus management logic with portaling currently does not handle usecases where there can be interactive elements between the domReference and floating elements.

Reproduction based off of the reusable Popover example: https://codesandbox.io/s/strange-bash-tlvzd9?file=/src/App.tsx. The button between trigger and content should represent such "hook higher in the component tree" approach, i.e. a complex scenario where the wrapper-trigger-content parts are not necessarily close together in the component tree.

I'm attempting to work on a fix to handle those in-between elements, but just in case pinging you in case you have some ideas or thoughts on this. Based on what I've understood so far, fixing this should also fix focus management issues with an external ref, as that scenario is very similar.

**Comment by mihkeleidast** [comment]
Alright, here's where I'm at currently: https://codesandbox.io/s/suspicious-johnson-z9tllr?file=/src/styles.css

This is my drawer implementation, so complete with the split-screen view on large screens, and "modal-y" behavior on smaller screens. I implemented the external store with jotai, but that was my personal preference so there wouldn't be a need for a component-specific global context provider. I suppose it could also be implemented with a context provider without much trouble.

Feel free to fork/edit/add to docs, if this seems useful!

Some notes:
1. This uses modal focus management always. I would still like to make it a non-modal dialog on large screens in split screen mode, to align the semantic behavior with the actual behavior (the rest of the page is accessible for mouse users). This is problematic due to focus management issues. See #2154 for more details.
2. Since this drawer is animated, I had to implement a "children caching" solution so the content would always be available during the "out" transition. Quite often the Drawers `open` state is controlled by some external global, and when it turns off, the content is underendered as well. This happens for example when the open state and content rendering is controlled by some route path logic. This does seem like a common issue TBH, but not sure if it can be generalized into some utility hook in floating-ui.
3. One additional caveat I ran into today was that for singleton mode, the toggle prop in useClick needs to be turned off, otherwise it will close on second button click instead of simply changing content. I did not figure any simple way how to make this work, it needs to travel from DrawerTrigger to `getReferenceProps` to `useClick` internals somehow? IDK, felt like this is not too big of an issue for me at the moment, so I disabled it, but would still be good to figure it out, if you have any ideas.
4. One additional note in the form of a question. While investigation 3 I noticed that useClick hook adds several other event handlers to the reference element as well (pointer, mouse, keyboard events). This got me wondering if I'd need to handle the keyboard handlers when setting my singleton content as well? Or will it still always end up in the onClick handler, so externally doing only that is enough? If it should also be handled externally, could it be made simpler in some way, as I know what I want to happen when some specific reference opens a floating element, no matter the specific input type. Any insight into this?

Hope I did not forget anything important now, getting late here.

@atomiks friendly ping for visibility :)

**Comment by mihkeleidast** [comment]
Alright, for the 3rd issue in my last comment, I found a workaround. Instead of using the `getReferenceProps` from the synced global state, I can use the interactions hooks inside of DrawerTrigger, which can then depend on the `current` prop and disable toggling as needed. The only remaining caveat with this workaround is that `useClick` et al hooks require `context` to be defined, but since we're syncing it through the global store, it's not defined on the first render... And since the hooks can't be called conditionally, it needs some sort of a mock default context object for the first pass. Not sure if this can be further improved in some way or not, will see.

https://codesandbox.io/s/heuristic-dhawan-d41q8v?file=/src/App.tsx

**Comment by WilliamIPark** [comment]
I'm also running into this. Would definitely be good for the API to support this without workarounds.

**Comment by atomiks** [comment]
I can't really think of a way to fix this easily, unless we register all the event handlers (and other attributes) on the reference `ref` directly, rather than use React event handler props. I guess that's possible, since I don't think we need the portal event bubbling behavior, but makes it a bit unwieldy to implement (forgoing the declarative DOM update model). This would only make `getFloatingProps` and `getItemProps` relevant for `useInteractions` then.

As far as I can tell, the "workarounds" currently are:

- Hoisting `useFloating` to the component where the reference is located so the JSX can receive the props. The data can be passed as props to the floating element component.
- Using a layout effect to synchronize `getReferenceProps` to a parent component and then spreading it that way.

I guess this is more of a React data flow problem, than anything else?

**Comment by atomiks** [comment]
I made this comment in the other thread but re-sharing it here. This is in essence a React data flow (or scoping) problem. You encounter it all the time in React outside of this library to build any type of reusable component. Common patterns when building components that work with the "internal" prop getters, include:

---

- **The compound component pattern**: all the re-usable examples on the docs do this. Both the reference element and floating element components/JSX are in scope, so data flow naturally works.

`Tooltip` contains the logic. Can be passed down via context, so everything is available in scope.

```js
<Tooltip>
  <TooltipTrigger>Trigger</TooltipTrigger>
  <TooltipContent>Tooltip</TooltipContent>
</Tooltip>
```

---

- **The `cloneElement` pattern**: your reference JSX is passed as a prop (usually `children`) and the `ref` and `getReferenceProps` can be attached (not recommended by React anymore due to the implicitness, but it's very convenient):

```js
const ref = useMergeRefs([refs.setReference, children.ref]);
return <>
  {cloneElement(children, getReferenceProps({ref}))}
  <div ref={refs.setFloating} style={floatingStyles}>{content}</div>
</>;
```

```js
<Tooltip content="Hello">
  <button>Trigger</button>
</Tooltip>
```

---

- **The render props pattern**: your reference JSX receives the props via a render prop on the component:

```js
<Tooltip renderTrigger={(getProps) => <button {...getProps()} />}>
  Content
</Tooltip>
```

```js
function Tooltip({renderTrigger, children}) {
  const {refs, floatingStyles} = useFloating();
  const {getReferenceProps} = useInteractions([...]);
  return <>
    {renderTrigger(getReferenceProps)}
    <div ref={refs.setFloating} style={floatingStyles}>{children}</div>
  </>;
}
```

---

- **The hoisting pattern**: hoist the `useFloating` call higher in the component tree to where the reference component/JSX is so that it's in scope, and pass data down to the floating component as props.

```js
function Parent() {
  const floatingProps = useFloating();
  return <>
    <button ref={floatingProps.refs.setReference}>Trigger</button>
    <Tooltip floatingProps={floatingProps}>
      Content
    </Tooltip>
  </>;
}
```

---

- **The sync pattern**: sync `getReferenceProps` to a parent with a layout effect, like:

```js
function Parent() {
  const [trigger, setTrigger] = useState(null);
  const [getReferenceProps, setGetReferenceProps] = useState();
  return <>
    <button ref={setTrigger} {...getReferenceProps?.()} />
    <Tooltip setGetReferenceProps={setGetReferenceProps} />
  </>;
}
```

```js
function Tooltip({setGetReferenceProps}) {
  const {getReferenceProps} = useInteractions([...]);
  useLayoutEffect(() => {
    setGetReferenceProps(() => getReferenceProps);
  }, [setGetReferenceProps, getReferenceProps]);
}
```

---

There may be more, or a dedicated way to help with data flow using the library. For example, `open` and `onOpenChange` can be external, as they are controlled by you. 

In this case, we want some kind of external mechanism for the props spreading. Let me know if you have any ideas!

Ariakit recently introduced a model called [Component Stores](https://ariakit.org/guide/component-stores) that looks quite unique. But it's the same I think — in this case both are available in scope, everything is contained within that hook.

**Comment by mpereira** [comment]
My use case was having an SVG element not rendered by React be a floating reference, so the above workarounds couldn't work.

What I ended up doing:
```js
const { context, floatingStyles, refs } = useFloating({
  elements: {
    reference: externalSvgElement,
  },
});

const hover = useHover(context);
const focus = useFocus(context);
const { getReferenceProps, getFloatingProps } = useInteractions([
  hover,
  focus,
]);

const reactToDomEventType = {
  onBlur: 'blur',
  onFocus: 'focus',
  onMouseLeave: 'mouseleave',
  onMouseMove: 'mousemove',
  onPointerDown: 'pointerdown',
  onPointerEnter: 'pointerenter',
};

useLayoutEffect(() => {
  for (const [eventType, eventHandler] of Object.entries(getReferenceProps())) {
    externalSvgElement.addEventListener(
      reactToDomEventType[eventType],
      eventHandler
    );
  }
}, []);

```

**Comment by atomiks** [comment]
@mpereira couple issues:

- Not every prop from `getReferenceProps` is a function/event handler
- It's subject to stale closures - make sure `getReferenceProps` is a dependency of the layout effect

This should work correctly:

```js
useLayoutEffect(() => {
  const entries = Object.entries(getReferenceProps());

  for (const [key, value] of entries) {
    if (typeof value === 'function') {
      el.addEventListener(reactToDomEventType[key], value);
    } else if (value) {
      el.setAttribute(key, value);
    } else {
      el.removeAttribute(key);
    }
  }

  return () => {
    for (const [key, value] of entries) {
      if (typeof value === 'function') {
        el.removeEventListener(reactToDomEventType[key], value);
      } else {
        el.removeAttribute(key);
      }
    }
  };
}, [getReferenceProps]);
```

- A third potential issue is that the props assume they are operating in the context of the React synthentic event system.  While it may work currently, it may break if one of the event handlers starts assuming that context.

**Comment by mihkeleidast** [comment]
An issue I ran into today with my external reference syncing solution: since I am recreating `useInteractions` for both the main component and the trigger, they ended up in different `FloatingTree` contexts. Since I was using `useDismiss` in both places, the one on the external trigger triggered an outside press close event when interacting with a nested floating component. Not sure if this has been broken in my components the entire time or something changed recently... I disabled `useDismiss` on the external trigger for now, as it does not seem necessary - since the hook mostly attaches global event listeners, not spreadable props (at least my usecase does not use the spreadable props it does return), it seems fine, but I'm not entirely sure.

Anyhow - is there a way to attach the external trigger to the same floating tree context as the main component? I would feel "safer" if that was possible. Maybe we could share the tree through `context` instead of accessing it through context directly in `useDismiss`?

**Comment by mayank99** [comment]
in the latest version (0.26.x), i had to add `nativeEvent` manually. this code seems to work:

```ts
React.useEffect(() => {
  if (!reference) return;

  /** e.g. onPointerDown --> pointerdown */
  const domEventName = (e: string) => e.toLowerCase().substring(2);

  const cleanupValues: Record<string, any> = {};

  Object.entries(getReferenceProps()).forEach(([key, value]) => {
    if (typeof value === 'function') {
      // manually add `.nativeEvent` because floating-ui needs it
      const patchedHandler = (event: Event) => {
        value({ ...event, nativeEvent: event });
      };
      reference.addEventListener(domEventName(key), patchedHandler);
      cleanupValues[key] = patchedHandler;
    } else if (value) {
      cleanupValues[key] = reference.getAttribute(key);
      reference.setAttribute(key, value);
    }
  });

  return () => {
    Object.entries(cleanupValues).forEach(([key, value]) => {
      if (typeof value === 'function') {
        reference.removeEventListener(domEventName(key), value);
      } else {
        reference.removeAttribute(key);
      }
    });
  };
}, [reference, getReferenceProps]);
```

**Comment by atomiks** [comment]
> in the latest version (0.26.x), i had to add nativeEvent manually. this code seems to work:

So this was right then  😬 - if you're using this then just be aware that it can break at any time 

>  A third potential issue is that the props assume they are operating in the context of the React synthentic event system. While it may work currently, it may break if one of the event handlers starts assuming that context.

**Comment by atomiks** [comment]
I've worked on supporting this without any hacks needed: https://github.com/floating-ui/floating-ui/pull/2890

**Reply by andreasvirkus** [reply]
Could we somehow update the #2128 issue? For the longest time I missed that it got converted into a discussion and wasn't aware of the `useFloatingRootContext` hook (also missing from the docs)
